Add init command that accepts a driver as a parameter and initializes the current directory to be compatible with that driver.

createdbysk opened the pull request:
contrib/initializers/
==============
* Add an example for initializers.

* Makefile
========
* Add contrib/initializers.
* contrib/intializers has a template sub-directory. Therefore, copy the entire directory recursively in the build and install steps.

* util.sh
=======
* Add util::get_plugin_path that contains the plugin_path code that was originally in plugin_mgr::load() to allow the driver initializer to find the location of the plugins.

cli.sh
======
* Add cli::init()
* Add the help for the init command.

plugin_mgr.sh
=============
* Source initializers.

main.sh
=======
* Use util::get_plugin_path()

initializer.sh
==============
* Add new type of plugin.

Commit:
Add init command that accepts a driver as a parameter and initializes a repository for that driver. (#1)

* util.sh
=======
* Add util::get_plugin_path that contains the plugin_path code that was originally in plugin_mgr::load() to allow the driver initializer to find the location of the plugins.
* Add util::split_and_index(), which splits a string at the given delimiter and returns the part at the given index.

cli.sh
======
* Add cli::init()
* Add the help for the init command.

plugin_mgr.sh
=============
* Source initializers.

main.sh
=======
* Use util::get_plugin_path()

initializer.sh
==============
* Add new type of plugin.

benschw commented on src/app/cli.sh:
```diff
@@ -122,6 +128,7 @@ cli::help() {
 	echo "    horde command [options]"
 	echo
 	echo "COMMANDS:"
+	echo "    init driver_name             use driver_name.initializer.sh to initialize this directory"
```
match help style used elsewhere
```suggestion
	echo "    init [driver_name]             use driver_name.initializer.sh to initialize this directory"
```

benschw commented on src/app/lib/initializer.sh:
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+
+init::run() {
```
This isn't totally intuitive, but my intention wasn't that each plugin type implements `run`, but that it implements the subcommands which it provides. (`horde up` is actually an alias for `horde run`)

```suggestion
initializer::init() {
```

benschw replied:
also, the plugin type should be a noun: you run `init` on your `initializer`

benschw commented on src/app/lib/util.sh:
```diff
@@ -15,3 +15,56 @@ util::func_exists() {
+# GIVEN input
+# AND HORDE_PLUGIN_PATH is set
+# THEN return $HORDE_PLUGIN_PATH
+util::get_plugin_path() {
```
I like this: good idea

benschw commented on src/app/lib/util.sh:
```diff
@@ -15,3 +15,56 @@ util::func_exists() {
+# Given parameters string, delimiter, index
+# splits string at delimiter and returns the part at
+# the zero-based index.
+util::split_and_index() {
```
what is this for?

benschw commented:
It would also be cool to build a `hello` example initializer in the contrib path to serve as an example

Commit:
Changes from review by benschw.

Rename init::run to initializer::init. (#2)
Add contrib/initializers/ as an example for initializers.
Remove util::_split_and_index().

benschw committed:
Update src/app/cli.sh

Accept benschw's suggestion for this change.

Co-Authored-By: satvidh <satvidh@users.noreply.github.com>
